### PR TITLE
Add SARIF support

### PR DESCRIPTION
**Pull request by shaopeng-gh**
Add support for output in SARIF format.

Hello! We are interested in adding support for output in the open-standard SARIF format to puppet-lint. SARIF support is required to integrate it in [GitHub code scanning](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning). Doing so would make it available to every repo in GitHub and could result in increase in user base.

More about SARIF here:
[What is SARIF?](https://github.com/microsoft/sarif-tutorials/blob/main/docs/1-Introduction.md#what-is-sarif)
[Why SARIF?](https://github.com/microsoft/sarif-tutorials/blob/main/docs/1-Introduction.md#why-sarif)

After the support for SARIF output is added to the tool and published a new version, we will work on creating a starter workflow to make it available as a GitHub code scanner.

### CODE REVIEW

**Comment by shaopeng-gh** [comment]
@yongyan-gh
@eddynaka

File: lib/puppet-lint/bin.rb
Comment (ekohl) [inline lib/puppet-lint/bin.rb:113]: Isn't this the same?
```suggestion
    sarif_file = File.read(File.join(__dir__, 'report', 'sarif.json'))
```

File: lib/puppet-lint/bin.rb
Comment (ekohl) [inline lib/puppet-lint/bin.rb:76]: If you're already collecting all problems in an array, you can remove this part here and properly serialize JSON at the end.

File: lib/puppet-lint/bin.rb
Comment (ekohl) [inline lib/puppet-lint/bin.rb:83]: So this would be obsolete

File: lib/puppet-lint/bin.rb
Comment (ekohl) [inline lib/puppet-lint/bin.rb:94]: And here would be the print of JSON

**Review by ekohl — COMMENTED** [review]
I've been wondering if puppet-lint shouldn't properly implement a `--format` option. In #34 I already added a start with this.

File: lib/puppet-lint/bin.rb
Comment (shaopeng-gh) [reply lib/puppet-lint/bin.rb]: fixed

File: lib/puppet-lint/bin.rb
Comment (shaopeng-gh) [reply lib/puppet-lint/bin.rb]: yes, fixed and it is better because it is now pretty print,
before change manual adding [ and , will not align well.

**Comment by shaopeng-gh** [comment]
@ekohl 

SARIF format need to have a description of the rule (beside the message that is very specific the the error found which we already have now), 
and a help uri if it is available.

I can see in each rule it do have this info but in the comments.
I will go ahead and modify each rule and also expose them in the object properties so I can get in the code,

let me know if that sounds good?

image

**Comment by shaopeng-gh** [comment]
> @ekohl
> 
> SARIF format need to have a description of the rule (beside the message that is very specific the the error found which we already have now), and a help uri if it is available.
> 
> I can see in each rule it do have this info but in the comments. I will go ahead and modify each rule and also expose them in the object properties so I can get in the code,
> 
> let me know if that sounds good?
> 
> image
> 
> the above example is not a good example, the message do not have any extra data so it looks like it can be directly used as description as well, but we have rules like this: :message => "indentation of => is not properly aligned (expected in column #{arrow_column[level_idx]}, but found it in column #{arrow_tok.column})",
> 
> the rule description we are looking for, is static for a rule.

I have done making this change and updated the PR.

fyi the new data will be visible in SARIF viewer UI like below:
image

File: lib/puppet-lint.rb
Comment (ekohl) [inline lib/puppet-lint.rb:183]: This could affect usage via Rake. I don't know if people do this, but there are effectively 2 entry points: the bin file which you already found and Rake:
https://github.com/puppetlabs/puppet-lint/blob/020143b705b023946739eb44e7c7d99fcd087527/lib/puppet-lint/tasks/puppet-lint.rb#L88-L107=

File: lib/puppet-lint/bin.rb
Comment (ekohl) [inline lib/puppet-lint/bin.rb:98]: Should these 2 be mutually exclusive? Either JSON or sarif?

File: lib/puppet-lint/bin.rb
Comment (ekohl) [inline lib/puppet-lint/bin.rb:102]: There's also `delete_if` which is probably cleaner. Or you could be explicit and use [`Hash.slice`](https://ruby-doc.org/core-2.5.1/Hash.html#method-i-slice) and select the entries that are relevant.

**Review by ekohl — COMMENTED** [review]
What about plugins which haven't added a description yet? Will that work well? Should it also document something for plugin writers?

File: lib/puppet-lint.rb
Comment (shaopeng-gh) [reply lib/puppet-lint.rb:183]: thanks for all the inputs! I will set the PR as draft for now and look into them.

File: lib/puppet-lint/bin.rb
Comment (shaopeng-gh) [reply lib/puppet-lint/bin.rb]: yes, changed to use `delete_if`

**Comment by shaopeng-gh** [comment]
1. I have verified running with plugins don't have those 2 new properties, will not throw any error and will just have those 2 sarif properties not populated.
2. I found this file `lib/puppet-lint/checkplugin.rb` has some description, I added the 2 new properties.

---
In reply to: [930112225](https://github.com/puppetlabs/puppet-lint/pull/40#issuecomment-930112225) [](http://example.com/codeflow?ancestors=930112225,930112225)

File: .github/workflows/scan.yml
Comment (ekohl) [inline .github/workflows/scan.yml:1]: This is an interesting thing. Do you have an example of where this ran so I can see the resulting report?

File: lib/puppet-lint/bin.rb
Comment (ekohl) [inline lib/puppet-lint/bin.rb:101]: I'm not sure the description must be excluded. The help URI also doesn't hurt IMHO. Perhaps a maintainer can weigh in on that.

**Review by ekohl — COMMENTED** [review]
I think my concerns have been addressed. However, I'm not a maintainer on this repository.

@binford2k you may want to give some visibility on this. It could be interesting for developers.

File: .github/workflows/scan.yml
Comment (shaopeng-gh) [reply .github/workflows/scan.yml:1]: This is just a sample (or sometimes called a starter workflow) that the user of the puppet-lint can add to their own repo, after we got all things done. User can use it to scan their repo and get code scanning alerts of the .pp files in their repo.
GitHub support this and the format used is SARIF format.

this is a run in my fork as example: (it is not intended to run against the puppet-lint repo itself, but the user's repo)
https://github.com/shaopeng-gh/puppet-lint/security/code-scanning

for people don't have permission:
image

File: lib/puppet-lint/bin.rb
Comment (shaopeng-gh) [reply lib/puppet-lint/bin.rb:101]: Perhaps a maintainer can weigh in on that. ---- yes the only reason for this code block is I am trying to maintain the same JSON output as before, maintainer can advise if also want these 2 new properties in JSON and let me know to change accordingly.

File: .github/workflows/scan.yml
Comment (shaopeng-gh) [reply .github/workflows/scan.yml]: Matrix is not needed, I have removed it.

**Comment by ekohl** [comment]
At this point I think this looks good, but someone who actually has access on this module needs to step in and do a review. I can't even approve CI to run.

**Comment by alexjfisher** [comment]
@binford2k Can I nominate @ekohl for trusted contributor access to this repo? :)

**Comment by chelnak** [comment]
I've kicked off the CI run 👍

**Comment by shaopeng-gh** [comment]
Thanks for looking and triggering the CI! let me know.

(The plan is after the support for SARIF output is added to the tool, I will work on creating a starter workflow in the GitHub code scanner starter workflow repo to make it available as a GitHub code scanner. I have added more information in the description of this PR at the top. )

**Comment by chelnak** [comment]
@shaopeng-gh This looks fine and it looks like you have addressed all of the comments from @ekohl.

My only question is do we need the scan.yml in this repo? If it is an example, would it be better to include in in the README?

Thanks!

**Comment by shaopeng-gh** [comment]
@chelnak your are correct, I have removed this file form the pr. It is just an example I will be submitting it to GitHub code scanner starter workflow repo, and user will be able to use it. 
After that is done we can add a few lines in the readme mentioning there is a GitHub starter workflows available.

**Comment by shaopeng-gh** [comment]
A gentle ping, after this is merged and a new version created I will work on the GitHub code scanner starter workflow. Thanks!

**Review by david22swan — APPROVED** [review]
LGTM
Just need's a squash merge

**Comment by david22swan** [comment]
@shaopeng-gh Thanks for putting in the work :)

**Comment by chelnak** [comment]
@shaopeng-gh I'm really looking forward to seeing what happens with this.

Thank you 🙏

**Comment by shaopeng-gh** [comment]
@chelnak 
thanks, not familiar with how Ruby Gems are updated, can a new version to be published so that in the workflow I can use below command to install and have the --sarif support:
`gem install puppet-lint`
https://rubygems.org/gems/puppet-lint/versions I see last version is [2.5.2](https://rubygems.org/gems/puppet-lint/versions/2.5.2) - September 14, 2021

**Pull request merged**
